### PR TITLE
ES6 classes

### PR DESCRIPTION
**Pull request by arv**

### CODE REVIEW

**Comment by arv** [comment]
I’m going to apply this to my Quip checkout and make sure things work before continuing with mixins.

We should also update the readme to explain how this works and how it depends on Quip internal implementation for mixins etc.

File: src/info/persistent/react/jscomp/ReactCompilerPass.java [hunk 1]
```diff
@@ -1647,63 +1646,6 @@ private void visitClassExtendsReactComponent(
...
-   *     constructor(props) {
-   *       super(props);
-   *       ...
-   *       \/** @private {{...}} *\/
-   *       this.state = {...}; <== Find this statement node
-   *       ...
-   *     }
-   *   }
-   */
-  private Node findThisStateNode(Node constructorNode) {
```
Comment (arv) [inline src/info/persistent/react/jscomp/ReactCompilerPass.java]: We could also support both in React Closure Compiler

File: src/info/persistent/react/jscomp/ReactCompilerPass.java [hunk 2]
```diff
@@ -1203,6 +1421,299 @@ private static boolean isReactPropTypes(Node value) {
...
+  }
+
+  /**
+   * Finds the declaring the state type inside a constructor
+   * 
+   *   class C extends React.Component {
+   *     constructor(props) {
+   *       super(props);
+   *       ...
+   *       \/** @private {{...}} *\/
```
Comment (arv) [inline src/info/persistent/react/jscomp/ReactCompilerPass.java]: Should have been `@type`

File: src/info/persistent/react/jscomp/ReactCompilerPass.java [hunk 3]
```diff
@@ -1203,6 +1434,232 @@ private static boolean isReactPropTypes(Node value) {
...
+    // "Clear" the @export bit, otherwise the compiler will complain about its
+    // presence on module-scoped variables.
+    if (isExportedType && scope.isModuleScope()) {
+      JSDocInfoAccessor.setJSDocExport(jsDocInfo, false);
+    }
+
+    JSDocInfoBuilder jsDocBuilder;
+
+    // Do not add "<type name>Interface" since we can use the type directly
+
```
Comment (mihaip) [inline src/info/persistent/react/jscomp/ReactCompilerPass.java]: Nit: empty line?

File: src/info/persistent/react/jscomp/ReactCompilerPass.java [hunk 4]
```diff
@@ -324,9 +405,148 @@ public void visit(NodeTraversal t, Node n, Node parent) {
...
       // more efficiently done in one function intead of two.
     } else if (isReactCreateElement(n)) {
-      visitReactCreateElement(t, n);
+      // We have to defer dealing with this until we are done with the script
+      // because the propTypes and defaultProps might be out of line.
+      reactCreateElementNodes.add(new NodeAndScope(n, scope));
     } else if (isReactPropTypes(n)) {
       visitReactPropTypes(n);
+    } else if (isClassExtendsReactComponent(n)) {
+      visitClassExtendsReactComponent(t, n, reactClassInterfacePrototypePropsByName);
```
Comment (mihaip) [inline src/info/persistent/react/jscomp/ReactCompilerPass.java]: Where is `reactClassInterfacePrototypePropsByName` used?

File: src/info/persistent/react/jscomp/ReactCompilerPass.java [hunk 5]
```diff
@@ -1203,6 +1434,232 @@ private static boolean isReactPropTypes(Node value) {
...
+    boolean isExportedType = jsDocInfo != null && jsDocInfo.isExport();
+    // "Clear" the @export bit, otherwise the compiler will complain about its
+    // presence on module-scoped variables.
+    if (isExportedType && scope.isModuleScope()) {
+      JSDocInfoAccessor.setJSDocExport(jsDocInfo, false);
+    }
+
+    JSDocInfoBuilder jsDocBuilder;
+
+    // Do not add "<type name>Interface" since we can use the type directly
```
Comment (mihaip) [inline src/info/persistent/react/jscomp/ReactCompilerPass.java]: Can you add an overview somewhere how the interface methods are used/how mixins are handled? I know we talked about it, but I’m having trouble following the code/remembering.

File: src/info/persistent/react/jscomp/types.js [hunk 6]
```diff
@@ -560,14 +560,92 @@ function ReactModule() {}
...
+ReactModule.prototype.Component.prototype.state;
+
+/** @override */
+ReactModule.prototype.Component.prototype.context;
+
+/** @override */
+ReactModule.prototype.Component.prototype.refs;
+
+/** @override */
+ReactModule.prototype.Component.prototype.getInitialState;
```
Comment (mihaip) [inline src/info/persistent/react/jscomp/types.js]: Should this be marked as deprecated?

File: test/info/persistent/react/jscomp/ReactCompilerPassTest.java [hunk 7]
```diff
@@ -851,6 +1230,136 @@
...
+        "}" +
+        "componentWillUpdate(nextProps, nextState) {" +
+          "nextState.enabled.toFixed(2);" +
+        "}" +
+        "render() {" +
+          "return null;" +
+        "}" +
+      "}",
+      "JSC_INEXISTENT_PROPERTY");
+    // // Mixin methods that take state parameters are checked
```
Comment (mihaip) [inline test/info/persistent/react/jscomp/ReactCompilerPassTest.java]: Missing a TODO?

File: test/info/persistent/react/jscomp/ReactCompilerPassTest.java [hunk 8]
```diff
@@ -1006,6 +1533,51 @@
...
+    ReactCompilerPass.Options passOptions =
+        new ReactCompilerPass.Options();
+    passOptions.optimizeForSize = true;
+    passOptions.propTypesTypeChecking = true;
+    // - propTypes should get stripped
+    // - React.createMixin() calls should be inlined with just the spec
+    // - React.createClass and React.createElement calls should be replaced with
+    //   React$createClass and React$createElement aliases (which can get fully
+    //   renamed).
+    // test(
```
Comment (mihaip) [inline test/info/persistent/react/jscomp/ReactCompilerPassTest.java]: Remove?

File: test/info/persistent/react/jscomp/ReactCompilerPassTest.java [hunk 9]
```diff
@@ -1508,6 +2294,41 @@
...
+          "}\n" +
+          "render() {return null;}" +
+        "}\n" +
+        "Comp.propTypes = {" +
+          "numberProp: React.PropTypes.number.isRequired" +
+        "};\n" +
+        "React.createElement(Comp, {numberProp: 1});",
+        "JSC_NOT_FUNCTION_TYPE");
+    // // As do abstract mixin methods that use ReactProps as the type.
+    // testError(
```
Comment (mihaip) [inline test/info/persistent/react/jscomp/ReactCompilerPassTest.java]: This too?

File: src/info/persistent/react/jscomp/ReactCompilerPass.java (hunk 5 above)
Comment (arv) [reply src/info/persistent/react/jscomp/ReactCompilerPass.java]: Mixins are not there yet.

No interface is used (except in externs)

**Comment by arv** [comment]
PTAL

File: src/info/persistent/react/jscomp/ReactCompilerPass.java (hunk 5 above)
Comment (mihaip) [reply src/info/persistent/react/jscomp/ReactCompilerPass.java]: Maybe in that case references to “interface” should be removed?

**Comment by mihaip** [comment]
LGTM

File: src/info/persistent/react/jscomp/ReactCompilerPass.java (hunk 5 above)
Comment (arv) [reply src/info/persistent/react/jscomp/ReactCompilerPass.java]: I thought I did remove it?

**Comment by arv** [comment]
Manually committed to master as b4bb529ed7df1cdc3e3cee2255fb22816fde5ab3

**Pull request closed without merging**